crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,178 @@
+            b
+        } else if b == ScopedVisibilityConstraintId::ALWAYS_TRUE {
+            a
+        } else {
+            self.add(VisibilityConstraint::KleeneAnd(a, b))
+        }
+    }
+
+    /// Analyze the statically known visibility for a given visibility constraint.
+    pub(crate) fn evaluate(&self, db: &'db dyn Db, id: ScopedVisibilityConstraintId) -> Truthiness {
+        self.evaluate_impl(db, id, MAX_RECURSION_DEPTH)
+    }
+
+    fn evaluate_impl(
+        &self,
+        db: &'db dyn Db,
+        id: ScopedVisibilityConstraintId,
+        max_depth: usize,
+    ) -> Truthiness {
+        if max_depth == 0 {
+            return Truthiness::Ambiguous;
+        }
+
+        let visibility_constraint = &self.constraints[id];
+        match visibility_constraint {
+            VisibilityConstraint::AlwaysTrue => Truthiness::AlwaysTrue,
+            VisibilityConstraint::Ambiguous => Truthiness::Ambiguous,
+            VisibilityConstraint::VisibleIf(constraint) => Self::analyze_single(db, constraint),
+            VisibilityConstraint::VisibleIfNot(negated) => {
+                self.evaluate_impl(db, *negated, max_depth - 1).negate()
+            }
+            VisibilityConstraint::KleeneAnd(lhs, rhs) => {
+                let lhs = self.evaluate_impl(db, *lhs, max_depth - 1);
+
+                if lhs == Truthiness::AlwaysFalse {
+                    return Truthiness::AlwaysFalse;
+                }
+
+                let rhs = self.evaluate_impl(db, *rhs, max_depth - 1);
+
+                if rhs == Truthiness::AlwaysFalse {
+                    Truthiness::AlwaysFalse
+                } else if lhs == Truthiness::AlwaysTrue && rhs == Truthiness::AlwaysTrue {
+                    Truthiness::AlwaysTrue
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+            VisibilityConstraint::KleeneOr(lhs_id, rhs_id) => {
+                let lhs = self.evaluate_impl(db, *lhs_id, max_depth - 1);
+
+                if lhs == Truthiness::AlwaysTrue {
+                    return Truthiness::AlwaysTrue;
+                }
+
+                let rhs = self.evaluate_impl(db, *rhs_id, max_depth - 1);
+
+                if rhs == Truthiness::AlwaysTrue {
+                    Truthiness::AlwaysTrue
+                } else if lhs == Truthiness::AlwaysFalse && rhs == Truthiness::AlwaysFalse {
+                    Truthiness::AlwaysFalse
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+        }
+    }
+
+    fn analyze_single(db: &dyn Db, constraint: &Constraint) -> Truthiness {
+        match constraint.node {
+            ConstraintNode::Expression(test_expr) => {
+                let inference = infer_expression_types(db, test_expr);
+                let scope = test_expr.scope(db);
+                let ty =
+                    inference.expression_ty(test_expr.node_ref(db).scoped_expression_id(db, scope));
```

Comment:
Referencing a previous review comment by @MichaReiser: https://github.com/astral-sh/ruff/pull/14759#discussion_r1881630989 

This is probably still relevant here.

Reply:
@MichaReiser I discussed this with @carljm: Using `node_ref` here should be fine, since the `Expression`s always come from the same file from which we call `VisibilityConstraints::evaluate`.

Reply:
That might be, but the problem is that `evaluate` isn't a salsa query, nor is the code calling into `evaluate.` The problem is, to some extent, pre-existing because `symbol_id` already depends on the `UseDefMap`, that's why I think we can tackle it as a follow-up (but we definitely should): 

The flow I'm concerned about is:

* `Type::member` calls `types::symbol` which is not a salsa query (nor is `Type::member`)
* `symbol` calls `types::symbol_by_id` which resolves the use def map (so it's a pre-existing issue!)
* `symbol_by_id` calls into `declarations_ty` which calls `evaluate` and depends on the AST

There are other examples for the same flow which are all cross-module:

* `global_symbol`
* `Class:own_member`

We should add a comment if, for some reason, the constraint holds that `symbol_by_id` is only called for the same file and never cross files.

Reply:
https://github.com/astral-sh/ruff/pull/15080 could address this and may also help with performance if it reduce (caches) the constraints that need solving?

Reply:
Looks like #15080 has been rolled directly into this PR. That seems like a robust and general fix. I guess when we have a tracked query that takes >1 argument (and one of them is not a Salsa ID), then effectively under the hood Salsa will end up creating an ingredient key for this `(ScopeId, ScopedSymbolId)` pair?